iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp

```diff
@@ -893,7 +893,13 @@ Optional<SmallVector<int64_t, 4>> getOpNativeVectorSize<vector::TransferReadOp>(
   // Map to load4.
   auto rank = op.getVectorType().getRank();
   SmallVector<int64_t, 4> nativeSize(rank, 1);
-  nativeSize.back() = 4;
+  // Load 4 elements on the most inner dimension.
+  for (auto dim : llvm::enumerate(op.permutation_map().getResults())) {
+    if (auto dimExpr = dim.value().dyn_cast<AffineDimExpr>()) {
+      if (dimExpr.getPosition() == op.permutation_map().getNumDims() - 1)
+        nativeSize[dim.index()] = 4;
```

Comment:
The comment about lit tests was for this case. Making sure that when we vectorize with transposes involved, the correct dimension is vectorized...

Reply:
This is exercised by the matmul vectorization and we have lit test for that. With the new vectorization pushed by Nicolas the Matmul now has the transpose on the transfer read at the beginning and we unroll it like that and lower the transpose in a second step. Removing this code would break the matmul vectorization lit tests for instance. 
Does that make sense? Or do you think we should test it more explicitly?

Reply:
>That is a very specific pattern. I see why you need it though. Shouldnt this just live in core.

Yes I agree it should live in core as mentioned in the TODO. I expect a bit of discussion on the semantic of ShapeCastOp and so I would rather not be blocked by that at this point but I'm planning to discuss it with Nicolas and others Vector dialect owners once he is back.

Reply:
> This is exercised by the matmul vectorization and we have lit test for that. With the new vectorization pushed by Nicolas the Matmul now has the transpose on the transfer read at the beginning and we unroll it like that and lower the transpose in a second step. Removing this code would break the matmul vectorization lit tests for instance.
> Does that make sense? Or do you think we should test it more explicitly?

Awesome. Thats fine.